src/sema/eval.rs

```diff
@@ -196,3 +202,398 @@ pub fn eval_const_rational(
+                Ok(Expression::NumberLiteral(*loc, ty.clone(), n))
+            } else {
+                Ok(Expression::SignExt(*loc, ty.clone(), Box::new(expr)))
+            }
+        }
+
+        _ => Ok(expr.clone()),
+    }
+}
+
+/// Function that takes a BigInt and an expected type. If the number of bits in the type required to represent the BigInt is not suffiecient, it will return a diagnostic.
+fn overflow_check(result: &BigInt, ty: &Type, loc: &Loc) -> Option<Diagnostic> {
+    if let Type::Uint(bits) = ty {
+        // If the result sign is minus, throw an error.
+        if let Sign::Minus = result.sign() {
+            return Some(Diagnostic::error(
+                *loc,
+            format!( "negative value {} does not fit into type uint{}. Cannot implicitly convert signed literal to unsigned type.",result,ty.get_type_size()),
+            ));
+        }
+
+        // If bits of the result is more than bits of the type, throw and error.
+        if result.bits() > *bits as u64 {
+            return Some(Diagnostic::error(
+                *loc,
+                format!(
+                    "value {} does not fit into type uint{}.",
+                    result,
+                    ty.get_type_size(),
+                ),
+            ));
+        }
+    }
+
+    if let Type::Int(bits) = ty {
+        // If number of bits is more than what the type can hold. BigInt.bits() is not used here since it disregards the sign.
+        if result.to_signed_bytes_be().len() * 8 > (*bits as usize) {
+            return Some(Diagnostic::error(
+                *loc,
+                format!(
+                    "value {} does not fit into type int{}.",
+                    result,
+                    ty.get_type_size(),
+                ),
+            ));
+        }
+    }
+    None
+}
+
+/// Wrapper function for the overflow detection logic.
+/// It first calls a function that recurses a given expression and returns a list of arithmetic operations within the expression, and a list of diagnostics.
+/// Then, it calls overflow_check() on the retrieved arithmetic operations, and pushes the returned diagnostics to the namespace
+pub(super) fn verify_expression_for_overflow(expr: &Expression, ns: &mut Namespace) {
+    let recursed: &mut Vec<Expression> = &mut Vec::new();
+    let diagnostics: &mut Vec<Diagnostic> = &mut Vec::new();
+    let mut params = RecurseParams {
+        recursed,
+        diagnostics,
+    };
+
+    expr.recurse(&mut params, check_term_for_constant_overflow);
+
+    for iter in params.recursed {
+        if let Expression::NumberLiteral(loc, ty, res) = iter {
+            if let Some(diagnostic) = overflow_check(res, ty, loc) {
+                ns.diagnostics.push(diagnostic);
+            }
+        }
+    }
+    ns.diagnostics.append(params.diagnostics);
+
+    match eval_constants_in_expression(&expr) {
+        Ok(expression) => {
+            if let Expression::NumberLiteral(loc, ty, res) = &expression {
```

Comment:
If `eval_constant_in_expression()` return `Result<Option<Expression>, Diagnostic>` then these two lines can be replaced with `Ok(Some(expr)) => {`